File: Java/src/_0547_Number_of_Provinces/Solution.java
package _0547_Number_of_Provinces;

/**
 * Date: 2021/1/7 9:20
 * Content:
 * 执行用时：3 ms, 在所有 Java 提交中击败了30.46%的用户
 * 内存消耗：39.4 MB, 在所有 Java 提交中击败了53.03%的用户
 */
public class Solution {

    public static int[] p;

    private int find(int x) {
        if (x != p[x]) p[x] = find(p[x]);
        return p[x];
    }

    public int findCircleNum(int[][] isConnected) {

        int n = isConnected.length;
        p = new int[n];

        for (int i = 0; i < n; i++) p[i] = i;

        int res = n;
        for (int i = 0; i < n; i++)
            for (int j = i + 1; j < n; j++) {
                int p1 = find(i), p2 = find(j);
                if (isConnected[i][j] == 1 && p1 != p2) {
                    p[p1] = p2;
                    res--;
                }
            }

        return res;
    }

    public static void main(String[] args) {
        int[][] isConnected = {
                {1, 1, 0},
                {1, 1, 0},
                {0, 0, 1}
        };
//        int[][] isConnected = {
//                {1, 0, 0, 1},
//                {0, 1, 1, 0},
//                {0, 1, 1, 1},
//                {1, 0, 1, 1}
//        };
        System.out.println((new Solution()).findCircleNum(isConnected));
    }
}

File: Python/_0547_Number_of_Provinces/Solution.py
# coding=utf-8
# Date: 2021/6/6 11:17
from typing import List


# 执行用时：76 ms, 在所有 Python3 提交中击败了28.55%的用户
# 内存消耗：15.3 MB, 在所有 Python3 提交中击败了41.00%的用户
class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        p = [i for i in range(210)]  # 并查集

        n = len(isConnected)
        res = n
        for i in range(n):
            for j in range(i):
                p1 = self.find(p, i); p2 = self.find(p, j)
                if isConnected[i][j] == 1 and p1 != p2:
                    res -= 1
                    p[p1] = p2
        return res

    def find(self, p, x):
        if x != p[x]:
            p[x] = self.find(p, p[x])
        return p[x]


if __name__ == "__main__":
    isConnected = [[1, 1, 0], [1, 1, 0], [0, 0, 1]]
    print(Solution().findCircleNum(isConnected))  # 2
